File: apps/ai-worker/app/schemas.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


Confidence = Annotated[float, Field(ge=0, le=1)]
ContentType = Literal[
    "single_shop_visit",
    "multi_shop_visit",
    "city_food_collection",
    "travel_vlog_with_shops",
    "food_review_not_shop",
    "not_physical_shop",
    "non_shop_visit",
    "unknown",
]
RiskFlag = Literal[
    "non_shop_visit_possible",
    "shop_name_missing",
    "shop_name_ambiguous",
    "generic_name_risk",
    "multiple_shops_in_video",
    "address_missing",
    "city_missing",
    "poi_no_candidate",
    "poi_low_confidence",
    "poi_many_same_name_candidates",
    "chain_store_branch_uncertain",
    "closed_or_moved_mentioned",
    "comment_conflict",
    "asr_low_quality",
    "subtitle_missing",
    "insufficient_evidence",
    "ai_output_incomplete",
    "needs_manual_review",
]
MissingField = Literal[
    "shop_name",
    "city",
    "district",
    "business_area",
    "exact_address",
    "poi",
    "opening_hours",
    "phone",
    "recommended_dishes",
    "avoid_points",
    "service",
    "environment",
    "queue",
    "parking",
    "reservation",
]
Sentiment = Literal["positive", "neutral", "negative", "mixed", "controversial", "unknown"]


class TranscriptSegment(StrictModel):
    segment_id: str | None = None
    start_sec: float
    end_sec: float
    text: str
    confidence: float | None = None


class VideoMetadata(StrictModel):
    video_id: str
    bvid: str
    creator_id: str
    title: str
    description: str | None = None
    tags: list[str] = Field(default_factory=list)
    category: str | None = None
    evidence: list["MetadataEvidence"] = Field(default_factory=list)


class MetadataEvidence(StrictModel):
    evidence_id: str
    source: Literal["title", "description", "tag"]
    text: str


class VideoAnalysisRequest(StrictModel):
    video_metadata: VideoMetadata
    transcript_segments: list[TranscriptSegment] = Field(default_factory=list)
    comment_samples: list["CommentSample"] = Field(default_factory=list)
    comment_signals: dict = Field(default_factory=dict)
    previous_stage_outputs: dict = Field(default_factory=dict)


class AsrResponse(StrictModel):
    source: Literal["asr"]
    language: str = "zh-CN"
    model_provider: str = "groq"
    model_name: str = "whisper-large-v3-turbo"
    content_text: str
    segments: list[TranscriptSegment]


class VideoClassificationResponse(StrictModel):
    schema_version: Literal["video_classification.v1"] = "video_classification.v1"
    video_id: str
    bvid: str
    is_shop_visit: bool
    content_type: ContentType
    confidence: float
    primary_city_hints: list[str] = Field(default_factory=list)
    primary_category_hints: list[str] = Field(default_factory=list)
    reason_codes: list[str] = Field(default_factory=list)
    risk_flags: list[RiskFlag] = Field(default_factory=list)
    need_manual_review: bool = False
    evidence_ids: list[str] = Field(default_factory=list)


class CommentSample(StrictModel):
    comment_id: str
    content: str
    like_count: int | None = None
    reply_count: int | None = None
    sample_type: str | None = None
    contains_location_signal: bool = False
    contains_shop_signal: bool = False


class LocationQuestion(StrictModel):
    text_summary: str
    count: int = Field(ge=0)
    evidence_ids: list[str] = Field(default_factory=list)


class ShopNameMention(StrictModel):
    candidate_name: str
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class AddressMention(StrictModel):
    text: str
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class AspectSentiment(StrictModel):
    sentiment: Sentiment
    summary: str
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class CommentSignalResponse(StrictModel):
    schema_version: Literal["comment_signal.v1"] = "comment_signal.v1"
    video_id: str
    sample_strategy: "SampleStrategy" = Field(default_factory=lambda: SampleStrategy())
    location_questions: list[LocationQuestion] = Field(default_factory=list)
    shop_name_mentions: list[ShopNameMention] = Field(default_factory=list)
    address_mentions: list[AddressMention] = Field(default_factory=list)
    status_mentions: list[dict[str, Any]] = Field(default_factory=list)
    aspect_sentiments: dict[str, AspectSentiment] = Field(default_factory=dict)
    risk_flags: list[RiskFlag] = Field(default_factory=list)

class CategoryPayload(StrictModel):
    primary: str | None = None
    secondary: str | None = None
    confidence: Confidence


class SampleStrategy(StrictModel):
    hot_comments_count: int = Field(default=0, ge=0)
    latest_comments_count: int = Field(default=0, ge=0)
    keyword_comments_count: int = Field(default=0, ge=0)


class LocationHints(StrictModel):
    country: str | None = None
    province: str | None = None
    city: str | None = None
    district: str | None = None
    business_area: str | None = None
    address_text: str | None = None
    landmarks: list[str] = Field(default_factory=list)
    confidence: Confidence


class TimeRange(StrictModel):
    start_sec: float | None = None
    end_sec: float | None = None


class CardConclusion(StrictModel):
    name: str | None = None
    text: str | None = None
    reason: str | None = None
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class CardPayload(StrictModel):
    display_title: str
    subtitle: str | None = None
    recommend_reason: str
    recommendation_score: Confidence | None
    recommendation_score_evidence_ids: list[str] = Field(default_factory=list)
    cover_source: str | None = None
    tags: list[str] = Field(default_factory=list)
    recommended_dishes: list[CardConclusion] = Field(default_factory=list)
    avoid_points: list[CardConclusion] = Field(default_factory=list)
    suitable_scenes: list[str] = Field(default_factory=list)


class ReviewDimension(StrictModel):
    sentiment: Sentiment
    summary: str
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class CommentSummary(StrictModel):
    positive_points: list[str] = Field(default_factory=list)
    negative_points: list[str] = Field(default_factory=list)
    controversial_points: list[str] = Field(default_factory=list)
    recent_status_points: list[str] = Field(default_factory=list)
    confidence: Confidence
    evidence_ids: list[str] = Field(default_factory=list)


class StructuredVideoPayload(StrictModel):
    video_id: str
    bvid: str
    creator_id: str
    title: str
    content_type: ContentType
    is_shop_visit: bool
    overall_summary: str
    primary_city: str | None = None
    primary_categories: list[str] = Field(default_factory=list)
    analysis_confidence: Confidence
    risk_flags: list[RiskFlag] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)


class StructuredShopCandidate(StrictModel):
    candidate_id: str
    candidate_name: str | None
    normalized_name: str | None
    name_confidence: Confidence
    alias_names: list[str] = Field(default_factory=list)
    candidate_type: Literal["physical_shop", "unknown", "not_shop"]
    category: CategoryPayload
    location_hints: LocationHints
    time_range: TimeRange | None = None
    card_payload: CardPayload
    review_dimensions: dict[str, ReviewDimension] = Field(default_factory=dict)
    comment_summary: CommentSummary
    missing_fields: list[MissingField] = Field(default_factory=list)
    risk_flags: list[RiskFlag] = Field(default_factory=list)
    manual_review_reasons: list[str] = Field(default_factory=list)


class VideoStructuredAnalysisResponse(StrictModel):
    schema_version: Literal["video_structured_analysis.v2"] = "video_structured_analysis.v2"
    video: StructuredVideoPayload
    shop_candidates: list[StructuredShopCandidate] = Field(default_factory=list)


class RelevantCommentFilterResponse(StrictModel):
    relevant_comment_ids: list[str] = Field(default_factory=list)


class TranscriptFactResponse(StrictModel):
    candidate_name: str | None = None
    name_confidence: Confidence = 0.0
    name_evidence_ids: list[str] = Field(default_factory=list)
    location_hints: LocationHints
    category: CategoryPayload


class TranscriptOpinionResponse(StrictModel):
    attitude: Literal["recommend", "conditional", "not_recommend", "unclear"]
    recommend_reason: str
    recommendation_score: Confidence | None
    recommendation_score_evidence_ids: list[str] = Field(default_factory=list)
    recommended_dishes: list[CardConclusion] = Field(default_factory=list)
    avoid_points: list[CardConclusion] = Field(default_factory=list)


class AiCallTrace(StrictModel):
    call_index: int = Field(ge=0)
    stage: str
    provider: str = "minimax"
    model: str
    prompt_version: str
    input_hash: str
    input_payload: dict[str, Any]
    output_payload: dict[str, Any] | None = None
    raw_output_text: str | None = None
    usage: dict[str, Any] = Field(default_factory=dict)
    status: Literal["success", "failed", "invalid_json", "schema_error"]
    error_message: str | None = None


class AiResponseEnvelope(StrictModel):
    output: dict[str, Any]
    provider: str = "minimax"
    model: str
    prompt_version: str
    usage: dict[str, Any] = Field(default_factory=dict)
    raw_output_text: str | None = None
    subcalls: list[AiCallTrace] = Field(default_factory=list)

File: packages/shared/src/schemas.ts
import { z } from "zod";
import {
  contentTypes,
  evidenceSources,
  missingFields,
  poiMatchStatuses,
  riskFlags,
  sentiments,
} from "./enums";

export const uuidSchema = z.string().uuid();
export const confidenceSchema = z.number().min(0).max(1);

export const pipelineRunTypes = [
  "creator_video_sync",
  "creator_profile_sync",
  "bilibili_auth_check",
  "video_processing",
  "video_asr_retry",
  "video_ai_retry",
  "poi_match",
] as const;

export const taskAcceptedResponseSchema = z.object({
  run_id: uuidSchema,
  job_id: uuidSchema.nullable(),
  run_type: z.enum(pipelineRunTypes),
  entity_type: z.string(),
  entity_id: uuidSchema,
  status: z.enum(["queued", "running"]),
});

export const adminTaskEventSchema = z.object({
  type: z.enum(["run.created", "run.updated", "pipeline.event"]),
  run_id: uuidSchema,
  run_type: z.enum(pipelineRunTypes).optional(),
  entity_type: z.string(),
  entity_id: uuidSchema,
  status: z.enum(["queued", "running", "success", "failed", "cancelled"]).optional(),
  event_id: uuidSchema.optional(),
  stage: z.string().optional(),
  event_type: z.string().optional(),
  level: z.string().optional(),
  progress_percent: z.union([z.number(), z.string(), z.null()]).optional(),
  created_at: z.string().optional(),
  updated_at: z.string().optional(),
});

export const evidenceSchema = z.object({
  id: z.string(),
  source: z.enum(evidenceSources),
  source_id: z.string().nullable().optional(),
  text: z.string().min(1),
  start_sec: z.number().nullable().optional(),
  end_sec: z.number().nullable().optional(),
  comment_id: z.string().nullable().optional(),
  confidence: confidenceSchema.nullable().optional(),
});

export const videoClassificationResultSchema = z.object({
  schema_version: z.literal("video_classification.v1"),
  video_id: z.string(),
  bvid: z.string(),
  is_shop_visit: z.boolean(),
  content_type: z.enum(contentTypes),
  confidence: confidenceSchema,
  primary_city_hints: z.array(z.string()).default([]),
  primary_category_hints: z.array(z.string()).default([]),
  reason_codes: z.array(z.string()).default([]),
  risk_flags: z.array(z.enum(riskFlags)).default([]),
  need_manual_review: z.boolean(),
  evidence_ids: z.array(z.string()).default([]),
});

export const commentSignalExtractionSchema = z.object({
  schema_version: z.literal("comment_signal.v1"),
  video_id: z.string(),
  sample_strategy: z
    .object({
      hot_comments_count: z.number().int().nonnegative().default(0),
      latest_comments_count: z.number().int().nonnegative().default(0),
      keyword_comments_count: z.number().int().nonnegative().default(0),
    })
    .default({}),
  location_questions: z
    .array(
      z.object({
        text_summary: z.string(),
        count: z.number().int().nonnegative(),
        evidence_ids: z.array(z.string()).default([]),
      }),
    )
    .default([]),
  shop_name_mentions: z
    .array(
      z.object({
        candidate_name: z.string(),
        confidence: confidenceSchema,
        evidence_ids: z.array(z.string()).default([]),
      }),
    )
    .default([]),
  address_mentions: z
    .array(
      z.object({
        text: z.string(),
        confidence: confidenceSchema,
        evidence_ids: z.array(z.string()).default([]),
      }),
    )
    .default([]),
  status_mentions: z.array(z.record(z.unknown())).default([]),
  aspect_sentiments: z.record(
    z.object({
      sentiment: z.enum(sentiments),
      summary: z.string(),
      confidence: confidenceSchema,
      evidence_ids: z.array(z.string()).default([]),
    }),
  ).default({}),
  risk_flags: z.array(z.enum(riskFlags)).default([]),
});

const conclusionSchema = z.object({
  name: z.string().optional(),
  text: z.string().optional(),
  reason: z.string().optional(),
  confidence: confidenceSchema,
  evidence_ids: z.array(z.string()).default([]),
}).strict();

const reviewDimensionSchema = z.object({
  sentiment: z.enum(sentiments),
  summary: z.string(),
  confidence: confidenceSchema,
  evidence_ids: z.array(z.string()).default([]),
}).strict();

export const videoStructuredAnalysisSchema = z.object({
  schema_version: z.literal("video_structured_analysis.v1"),
  video: z.object({
    video_id: z.string(),
    bvid: z.string(),
    creator_id: z.string(),
    title: z.string(),
    content_type: z.enum(contentTypes),
    is_shop_visit: z.boolean(),
    overall_summary: z.string(),
    primary_city: z.string().nullable().optional(),
    primary_categories: z.array(z.string()).default([]),
    analysis_confidence: confidenceSchema,
    risk_flags: z.array(z.enum(riskFlags)).default([]),
    evidence_ids: z.array(z.string()).default([]),
  }).strict(),
  shop_candidates: z.array(
    z.object({
      candidate_id: z.string(),
      candidate_name: z.string().nullable(),
      normalized_name: z.string().nullable(),
      name_confidence: confidenceSchema,
      alias_names: z.array(z.string()).default([]),
      candidate_type: z.enum(["physical_shop", "unknown", "not_shop"]),
      category: z.object({
        primary: z.string().nullable(),
        secondary: z.string().nullable(),
        confidence: confidenceSchema,
      }).strict(),
      location_hints: z.object({
        country: z.string().nullable().optional(),
        province: z.string().nullable().optional(),
        city: z.string().nullable().optional(),
        district: z.string().nullable().optional(),
        business_area: z.string().nullable().optional(),
        address_text: z.string().nullable().optional(),
        landmarks: z.array(z.string()).default([]),
        confidence: confidenceSchema,
      }).strict(),
      time_range: z
        .object({
          start_sec: z.number().nullable().optional(),
          end_sec: z.number().nullable().optional(),
        }).strict()
        .nullable()
        .optional(),
      card_payload: z.object({
        display_title: z.string(),
        subtitle: z.string().nullable().optional(),
        recommend_reason: z.string(),
        avg_price_hint: z.string().nullable().optional(),
        cover_source: z.string().nullable().optional(),
        tags: z.array(z.string()).default([]),
        recommended_dishes: z.array(conclusionSchema).default([]),
        avoid_points: z.array(conclusionSchema).default([]),
        suitable_scenes: z.array(z.string()).default([]),
      }).strict(),
      review_dimensions: z.record(reviewDimensionSchema).default({}),
      comment_summary: z.object({
        positive_points: z.array(z.string()).default([]),
        negative_points: z.array(z.string()).default([]),
        controversial_points: z.array(z.string()).default([]),
        recent_status_points: z.array(z.string()).default([]),
        confidence: confidenceSchema,
        evidence_ids: z.array(z.string()).default([]),
      }).strict(),
      missing_fields: z.array(z.enum(missingFields)).default([]),
      risk_flags: z.array(z.enum(riskFlags)).default([]),
      manual_review_reasons: z.array(z.string()).default([]),
    }).strict(),
  ),
}).strict();

export const poiMatchResultSchema = z.object({
  schema_version: z.literal("poi_match.v1"),
  candidate_id: z.string(),
  provider: z.enum(["amap", "tencent", "baidu"]),
  selected_poi: z
    .object({
      provider_poi_id: z.string(),
      name: z.string(),
      address: z.string().nullable().optional(),
      province: z.string().nullable().optional(),
      city: z.string().nullable().optional(),
      district: z.string().nullable().optional(),
      business_area: z.string().nullable().optional(),
      location: z.object({
        lng: z.number(),
        lat: z.number(),
        coord_type: z.enum(["gcj02", "bd09", "wgs84"]),
      }),
      category: z.string().nullable().optional(),
      raw_provider_payload_id: z.string().nullable().optional(),
    })
    .nullable(),
  candidates: z.array(
    z.object({
      provider_poi_id: z.string(),
      name: z.string(),
      address: z.string().nullable().optional(),
      province: z.string().nullable().optional(),
      city: z.string().nullable().optional(),
      district: z.string().nullable().optional(),
      business_area: z.string().nullable().optional(),
      location: z.object({
        lng: z.number(),
        lat: z.number(),
        coord_type: z.enum(["gcj02", "bd09", "wgs84"]),
      }),
      category: z.string().nullable().optional(),
      category_code: z.string().nullable().optional(),
      match_features: z.record(z.number()).default({}),
      match_score: z.number().min(0).max(1),
    }),
  ).default([]),
  match_score: z.number().min(0).max(1),
  match_status: z.enum(poiMatchStatuses),
  risk_flags: z.array(z.enum(riskFlags)).default([]),
  manual_review_reasons: z.array(z.string()).default([]),
});

export const publishedShopSnapshotSchema = z.object({
  shop_id: z.string(),
  canonical_name: z.string(),
  display_name: z.string(),
  poi: z.object({
    provider: z.string(),
    provider_poi_id: z.string(),
    address: z.string().nullable().optional(),
    location: z.object({
      lng: z.number(),
      lat: z.number(),
      coord_type: z.enum(["gcj02", "bd09", "wgs84"]),
    }),
  }),
  category: z.object({
    primary: z.string().nullable().optional(),
    secondary: z.string().nullable().optional(),
  }),
  card: z.object({
    title: z.string(),
    subtitle: z.string().nullable().optional(),
    recommend_reason: z.string(),
    avg_price_hint: z.string().nullable().optional(),
    tags: z.array(z.string()).default([]),
    cover_url: z.string().url().nullable().optional(),
    source_creator_avatars: z.array(z.string()).default([]),
  }),
  aggregated_review: z.record(reviewDimensionSchema).default({}),
  source_stats: z.record(z.unknown()).default({}),
  quality: z.record(z.unknown()).default({}),
});

export const sourceVideoSchema = z.object({
  video_id: z.string().optional(),
  title: z.string(),
  source_url: z.string(),
  bvid: z.string().optional(),
  cover_url: z.string().nullable().optional(),
  creator_name: z.string().optional(),
});

export const mapShopSchema = z.object({
  id: z.string(),
  display_name: z.string(),
  city: z.string().nullable().optional(),
  district: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
  lng: z.union([z.number(), z.string()]).nullable().optional(),
  lat: z.union([z.number(), z.string()]).nullable().optional(),
  coord_type: z.enum(["gcj02", "bd09", "wgs84"]).optional(),
  card_payload: z.record(z.unknown()).nullable().optional(),
  quality: z.record(z.unknown()).nullable().optional(),
  source_videos: z.array(sourceVideoSchema).default([]),
});

export const shopSearchResultSchema = mapShopSchema.extend({
  rank_score: z.number().optional(),
});

export type VideoClassificationResult = z.infer<typeof videoClassificationResultSchema>;
export type CommentSignalExtraction = z.infer<typeof commentSignalExtractionSchema>;
export type VideoStructuredAnalysis = z.infer<typeof videoStructuredAnalysisSchema>;
export type PoiMatchResult = z.infer<typeof poiMatchResultSchema>;
export type PublishedShopSnapshot = z.infer<typeof publishedShopSnapshotSchema>;
export type MapShop = z.infer<typeof mapShopSchema>;
export type ShopSearchResult = z.infer<typeof shopSearchResultSchema>;
export type TaskAcceptedResponse = z.infer<typeof taskAcceptedResponseSchema>;
export type AdminTaskEvent = z.infer<typeof adminTaskEventSchema>;
